modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPreloader.java

```diff
@@ -351,25 +342,26 @@ private IgniteCheckedException stopError() {
     }
 
     /** {@inheritDoc} */
-    @Override public void handleSupplyMessage(int idx, UUID id, final GridDhtPartitionSupplyMessage s) {
-        if (!enterBusy())
-            return;
+    @Override public void handleSupplyMessage(UUID id, final GridDhtPartitionSupplyMessage s) {
+        if (demander.prepareSupplyMessage(s)) {
+            ctx.kernalContext().closure().runLocalSafe(() -> {
```

Comment:
We don't need `runLocalSafe` here, we should use `GridIoPolicy.REBALANCE_POOL` at the moment of message sending (for demander and supplier). 
Such a technique doesn't free the `PUBLIC_POOL` threads and when it will be a really busy rebalance messages will stuck on it.